[WIP] SimpleFSDP prototype frontend changes

yf225 opened the pull request:
This is built on top of @fmassa @wanchaol @tianyu-l 's great SimpleFSDP prototype.

On top of the above work, this PR added the following changes on API side, mostly to **make SimpleFSDP a drop-in replacement of FSDP2 for internal models**:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in internal workflow
- Class cache for SimpleFSDPModule to avoid recompilation per module
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)

There are quite a few `NOTE(yf225)` inline notes to explain motivation of specific changes.

TODO:
- Distributed checkpointing verification
- Prefetching before first FSDP module
- UX: SimpleFSDP requires `active_parametrization()` ctx manager to toggle between sharded state and unsharded state. Optimizer init wants sharded state and training loop wants unsharded state. We might just ask user to always wrap their training loop with with `active_parametrization()`.
- Merge SimpleFSDPModule and FSDPModule into one class, once we are ready to use SimpleFSDP as compiled path of FSDP2 and hide it behind FSDP2 frontend API


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140360
* #140809



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

yf225 committed:
[WIP] SimpleFSDP prototype frontend revamp

[ghstack-poisoned]

yf225 committed:
Update on "[WIP] SimpleFSDP prototype frontend revamp"

This built on top of fmassa wanchaol tianyu-l 's great SimpleFSDP prototype.
On top of the above work, this PR added the following changes on API side:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in Ads workflow
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)






cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
> FSDP2 set_post_optim_event (expected by Ads workflow)

I do not think this will be needed since `set_post_optim_event` is only meant for implicit prefetching, which SimpleFSDP will not use.

yf225 committed:
Update on "[WIP] SimpleFSDP prototype frontend changes"

This is built on top of fmassa wanchaol tianyu-l 's great SimpleFSDP prototype.

On top of the above work, this PR added the following changes on API side, mostly to **make SimpleFSDP a drop-in replacement of FSDP2 for internal models**:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in internal workflow
- Class cache for SimpleFSDPModule to avoid recompilation per module
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)

There are quite a few `NOTE(yf225)` inline notes to explain motivation of specific changes.

TODO:
- Distributed checkpointing verification
- Prefetching before first FSDP module
- FSDP2 set_post_optim_event (expected by Ads workflow)
- UX: SimpleFSDP requires `active_parametrization()` ctx manager to toggle between sharded state and unsharded state. Optimizer init wants sharded state and training loop wants unsharded state. We might just ask user to always wrap their training loop with with `active_parametrization()`.





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yf225 committed:
Update on "[WIP] SimpleFSDP prototype frontend changes"

This is built on top of fmassa wanchaol tianyu-l 's great SimpleFSDP prototype.

On top of the above work, this PR added the following changes on API side, mostly to **make SimpleFSDP a drop-in replacement of FSDP2 for internal models**:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in internal workflow
- Class cache for SimpleFSDPModule to avoid recompilation per module
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)

There are quite a few `NOTE(yf225)` inline notes to explain motivation of specific changes.

TODO:
- Distributed checkpointing verification
- Prefetching before first FSDP module
- UX: SimpleFSDP requires `active_parametrization()` ctx manager to toggle between sharded state and unsharded state. Optimizer init wants sharded state and training loop wants unsharded state. We might just ask user to always wrap their training loop with with `active_parametrization()`.
- Merge SimpleFSDPModule and FSDPModule into one class, once we are ready to use SimpleFSDP as compiled path of FSDP2 and hide it behind FSDP2 frontend API





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yf225 committed:
Update on "[WIP] SimpleFSDP prototype frontend changes"

This is built on top of fmassa wanchaol tianyu-l 's great SimpleFSDP prototype.

On top of the above work, this PR added the following changes on API side, mostly to **make SimpleFSDP a drop-in replacement of FSDP2 for internal models**:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in internal workflow
- Class cache for SimpleFSDPModule to avoid recompilation per module
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)

There are quite a few `NOTE(yf225)` inline notes to explain motivation of specific changes.

TODO:
- Distributed checkpointing verification
- Prefetching before first FSDP module
- UX: SimpleFSDP requires `active_parametrization()` ctx manager to toggle between sharded state and unsharded state. Optimizer init wants sharded state and training loop wants unsharded state. We might just ask user to always wrap their training loop with with `active_parametrization()`.
- Merge SimpleFSDPModule and FSDPModule into one class, once we are ready to use SimpleFSDP as compiled path of FSDP2 and hide it behind FSDP2 frontend API





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yf225 committed:
Update on "[WIP] SimpleFSDP prototype frontend changes"

This is built on top of fmassa wanchaol tianyu-l 's great SimpleFSDP prototype.

On top of the above work, this PR added the following changes on API side, mostly to **make SimpleFSDP a drop-in replacement of FSDP2 for internal models**:
- Remove .parametrizations from parameter FQN (copied from Francisco's example)
- Module-level input/output dtype casting
- LazyModuleMixin handling
- Created SimpleFSDPModule class to mimic FSDPModule, for easier drop-in replacement in internal workflow
- Class cache for SimpleFSDPModule to avoid recompilation per module
- FSDP2-style nested wrapping -> verified in prototype
- Reconcile `ac_mode` and FSDP2 `reshard_after_forward` -> changed `ac_mode` to `reshard_after_forward` in this prototype, need further validation (esp. when used together with module-level or op-level AC)

There are quite a few `NOTE(yf225)` inline notes to explain motivation of specific changes.

TODO:
- Distributed checkpointing verification
- Prefetching before first FSDP module
- UX: SimpleFSDP requires `active_parametrization()` ctx manager to toggle between sharded state and unsharded state. Optimizer init wants sharded state and training loop wants unsharded state. We might just ask user to always wrap their training loop with with `active_parametrization()`.
- Merge SimpleFSDPModule and FSDPModule into one class, once we are ready to use SimpleFSDP as compiled path of FSDP2 and hide it behind FSDP2 frontend API





cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented:
> We might just ask user to always wrap their training loop with with active_parametrization().

sounds kind of bad, imo. can we do it some other way?

> Prefetching before first FSDP module

Any proposal for how the user Api would look for calling prefetch, in a way that is compiler-respected?  what if its prefetching for a non-module-aligned bucket?

ultranity commented:
Just found this great SimpleFSDP prototype.
Is SimpleFSDP a torch.compile enhancement to FSDP2, or a successor like FSDP2.5/3.0?
And is there any schedules on making it stable/merged?

awgu commented:
@ultranity I think SimpleFSDP is a fundamentally different approach to FSDP that is compile-first. The eager-mode implementation has no optimizations, and at least for now, without full graph, SimpleFSDP may not be able to get the best performance (e.g. cannot overlap collectives across graphs).

There is ongoing exploratory work to make SimpleFSDP more production-ready. However, SimpleFSDP may not be deployable to users who cannot ensure full graph compile. It would also offer a different UX for power users to customize it, as all optimizations are done in the compiler.
